File: sdk/typescript/src/attestation.ts
// Copyright 2026 AlphaOne LLC
// SPDX-License-Identifier: Apache-2.0

/**
 * Ed25519 agent-attestation signing for the HTTP store path (#626 / #2455).
 *
 * ## Why this module exists
 *
 * `POST /api/v1/memories` is classified `WriteSurface::HttpDirect`, which
 * **fails closed by default** (`src/identity/attest.rs:130-136` — the `_` arm
 * of `resolve_require_agent_attestation` returns `true` for `HttpDirect` when
 * the env var is unset). An unsigned HTTP store is answered with
 * `403 ATTESTATION_FAILED`.
 *
 * Before #2455 this SDK shipped no signing helper and `CreateMemoryRequest`
 * carried no `signature` / `created_at` fields, so `client.store()` could not
 * succeed against a stock daemon at all. This module closes that: it
 * reproduces the daemon's `SignableWrite` envelope byte-for-byte and signs it,
 * so the SDK works against the **shipped fail-closed default** rather than
 * requiring the operator to weaken the daemon with
 * `AI_MEMORY_REQUIRE_AGENT_ATTESTATION=0`.
 *
 * ## The envelope
 *
 * Ed25519 over the RFC 8949 4.2.1 deterministic-CBOR encoding of exactly seven
 * entries (`src/identity/sign.rs::canonical_cbor_write`):
 *
 * | key                | value                                              |
 * | ------------------ | -------------------------------------------------- |
 * | `_dst`             | domain tag, always `"ai-memory/write/v1"`           |
 * | `kind`             | memory kind (`"observation"`, `"decision"`, ...)    |
 * | `title`            | memory title                                        |
 * | `agent_id`         | the claiming agent id                               |
 * | `namespace`        | target namespace                                    |
 * | `created_at`       | RFC3339 timestamp the caller signed                 |
 * | `content_sha256`   | SHA-256 of content UTF-8 bytes (CBOR byte string)   |
 *
 * Map keys are emitted in encoded-key sort order (length first, then
 * bytewise), which for this fixed key set is the order shown. Byte-level
 * agreement with the Rust encoder is pinned by a fixture printed by the Rust
 * encoder itself — `sdk/fixtures/write_attestation_vector.json`, asserted in
 * `__tests__/attestation.test.ts`.
 *
 * ## Enrolling the key
 *
 * A signature only attests if the daemon has the matching public key bound to
 * the agent. Bind it before signing writes with
 * `client.bindAgentPubkey(agentId, signingKey)`. This issues an admin-gated
 * `POST /api/v1/agents/{id}/pubkey/challenge`, signs its transcript, and answers
 * `PUT /api/v1/agents/{id}/pubkey` with `{pubkey_b64, nonce, proof_b64}`.
 * A bare public key is insufficient.
 * Without a bound key the daemon cannot verify and still returns 403
 * (fail-closed, by design).
 *
 * ## Known limits (stated rather than hidden)
 *
 * - The daemon redacts secret-screened content *before* verifying
 *   (`identity::attest::redact_before_sign`). Content that trips the
 *   credential screen under `AI_MEMORY_SECRET_SCREEN_MODE=redact` verifies
 *   against different bytes than you signed and is refused; under the default
 *   `refuse` mode it is rejected earlier anyway. Do not put credentials in
 *   memories.
 * - `created_at` must be inside the daemon's +/-300s attestation freshness
 *   window (`ATTEST_CREATED_AT_SKEW_SECS`); sign shortly before sending.
 * - The signature commits to those six fields ONLY — not tags, priority, or
 *   metadata.
 *
 * Uses Node's built-in `node:crypto` — no third-party crypto dependency.
 */

import {
  createHash,
  createPrivateKey,
  createPublicKey,
  randomBytes,
  sign as cryptoSign,
  type KeyObject,
} from "node:crypto";

/** The `_dst` domain-separation value for the write envelope (#1931). */
export const DOMAIN_SEPARATION_TAG = "ai-memory/write/v1";

/** Length of an Ed25519 seed / public key in bytes. */
const KEY_LEN = 32;

// DER scaffolding for raw-seed <-> KeyObject conversion. Ed25519 PKCS#8 and
// SPKI encodings are fixed-length, so a constant prefix + the 32 raw bytes is
// the whole structure (RFC 8410).
const PKCS8_PREFIX = Buffer.from("302e020100300506032b657004220420", "hex");
const SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

// ---------------------------------------------------------------------------
// Deterministic CBOR (RFC 8949 4.2.1) — only the shapes this envelope needs.
// ---------------------------------------------------------------------------

function cborHead(major: number, length: number): Buffer {
  const prefix = major << 5;
  if (length < 24) return Buffer.from([prefix | length]);
  if (length < 0x100) return Buffer.from([prefix | 24, length]);
  if (length < 0x10000) {
    const b = Buffer.alloc(3);
    b[0] = prefix | 25;
    b.writeUInt16BE(length, 1);
    return b;
  }
  const b = Buffer.alloc(5);
  b[0] = prefix | 26;
  b.writeUInt32BE(length, 1);
  return b;
}

function cborText(value: string): Buffer {
  const raw = Buffer.from(value, "utf8");
  return Buffer.concat([cborHead(3, raw.length), raw]);
}

function cborBytes(value: Buffer): Buffer {
  return Buffer.concat([cborHead(2, value.length), value]);
}

/** SHA-256 over the UTF-8 bytes of `content` (the bounded commitment). */
export function contentSha256(content: string): Buffer {
  return createHash("sha256").update(content, "utf8").digest();
}

/** Fields the write signature commits to. */
export interface SignableWrite {
  agentId: string;
  namespace: string;
  title: string;
  /** Memory kind discriminant, e.g. `"observation"` / `"decision"`. */
  kind: string;
  /** RFC3339 timestamp; must match the `created_at` sent on the wire. */
  createdAt: string;
  /** 32-byte SHA-256 digest from {@link contentSha256}. */
  contentSha256: Buffer;
}

/**
 * Deterministic-CBOR encoding of the `SignableWrite` envelope.
 *
 * Byte-identical to `crate::identity::sign::canonical_cbor_write`.
 *
 * @throws {RangeError} if `contentSha256` is not exactly 32 bytes.
 */
export function canonicalCborWrite(write: SignableWrite): Buffer {
  if (write.contentSha256.length !== KEY_LEN) {
    throw new RangeError(
      `contentSha256 must be a 32-byte SHA-256 digest, got ${write.contentSha256.length}`,
    );
  }
  // RFC 8949 4.2.1 orders map keys by their ENCODED bytes: shorter keys first,
  // then bytewise. For this fixed key set that is the literal order below
  // ('_' 0x5F sorts before 'k' 0x6B at length 4). Hard-coding the resolved
  // order keeps the encoder simple and makes any future key addition an
  // explicit, reviewable re-ordering rather than a silent one. The order is
  // pinned against the Rust-emitted fixture.
  const entries: Array<[string, Buffer]> = [
    ["_dst", cborText(DOMAIN_SEPARATION_TAG)],
    ["kind", cborText(write.kind)],
    ["title", cborText(write.title)],
    ["agent_id", cborText(write.agentId)],
    ["namespace", cborText(write.namespace)],
    ["created_at", cborText(write.createdAt)],
    ["content_sha256", cborBytes(write.contentSha256)],
  ];
  const parts: Buffer[] = [cborHead(5, entries.length)];
  for (const [key, encodedValue] of entries) {
    parts.push(cborText(key), encodedValue);
  }
  return Buffer.concat(parts);
}

/** RFC3339 date-time, decomposed so the fraction survives `Date`'s millisecond floor. */
const RFC3339_PATTERN =
  /^(\d{4})-(\d{2})-(\d{2})[Tt ](\d{2}):(\d{2}):(\d{2})(?:\.(\d+))?([Zz]|[+-]\d{2}:\d{2})$/;

/**
 * Canonicalize an RFC3339 timestamp into the ONE `created_at` rendering both
 * daemon storage backends return byte-for-byte (#3422).
 *
 * The daemon REFUSES to attest any other rendering. `created_at` is inside the
 * signed `SignableWrite` envelope as TEXT, and every re-verification (the store
 * gate, and the federation receive path on a relayed row) rebuilds the envelope
 * from the PERSISTED row: SQLite returns the stored TEXT verbatim, while
 * PostgreSQL stores `TIMESTAMPTZ` (microseconds) and re-renders the readback as
 * `+00:00`. A signature minted over `"…Z"`, over a non-UTC offset, or over
 * nanosecond precision could therefore never be re-derived from a Postgres row.
 *
 * Canonical form: UTC, `+00:00` offset, microseconds TRUNCATED (never rounded —
 * that is what `TIMESTAMPTZ` does), and 0, 3 or 6 fractional digits — the
 * shortest width that represents the value exactly, matching chrono's
 * `SecondsFormat::AutoSi`.
 *
 * @throws {RangeError} if `value` is not an RFC3339 date-time.
 */
export function canonicalizeCreatedAt(value: string): string {
  const m = RFC3339_PATTERN.exec(value.trim());
  if (m === null) {
    throw new RangeError(`created_at must be an RFC3339 date-time, got: ${value}`);
  }
  const [, year, month, day, hour, minute, second, fraction = "", offset] = m;
  // TIMESTAMPTZ keeps microseconds and drops the rest — truncate, never round.
  const micros = Number(`${fraction}000000`.slice(0, 6));
  let offsetMinutes = 0;
  if (offset !== "Z" && offset !== "z") {
    const sign = offset.startsWith("-") ? -1 : 1;
    offsetMinutes = sign * (Number(offset.slice(1, 3)) * 60 + Number(offset.slice(4, 6)));
  }
  // Build via the UTC setters rather than `Date.UTC`, whose two-digit-year
  // legacy mapping would turn year 0099 into 1999.
  const moment = new Date(0);
  moment.setUTCFullYear(Number(year), Number(month) - 1, Number(day));
  moment.setUTCHours(Number(hour), Number(minute), Number(second), 0);
  const whole = new Date(moment.getTime() - offsetMinutes * 60_000)
    .toISOString()
    .slice(0, 19);
  let frac = "";
  if (micros !== 0) {
    frac =
      micros % 1000 === 0
        ? `.${String(micros / 1000).padStart(3, "0")}`
        : `.${String(micros).padStart(6, "0")}`;
  }
  return `${whole}${frac}+00:00`;
}

/**
 * Current UTC time in the canonical `created_at` form the daemon persists and
 * re-derives signatures from (#3422) — UTC, `+00:00`, microsecond-truncated.
 *
 * (`Date#toISOString` emits `Z` with a fixed 3-digit fraction; both the suffix
 * and a `.000` fraction are rejected by the daemon's attestation funnel, since
 * neither survives a PostgreSQL `TIMESTAMPTZ` round-trip unchanged.)
 */
export function rfc3339Now(date: Date = new Date()): string {
  return canonicalizeCreatedAt(date.toISOString());
}

// ---------------------------------------------------------------------------
// Key handling
// ---------------------------------------------------------------------------

function decodeBase64Any(value: string): Buffer {
  const raw = value.trim();
  // Buffer's base64 decoder accepts both alphabets and tolerates missing pad.
  return Buffer.from(raw, "base64");
}

/**
 * An Ed25519 signing key for agent write attestation.
 *
 * Wraps a 32-byte seed — the exact on-disk format the daemon's own key store
 * uses (`<key_dir>/<agent_id>.priv`, 32 raw bytes, mode 0600) — so a key made
 * by `ai-memory identity generate` loads directly with {@link fromSeed}.
 */
export class AgentSigningKey {
  private readonly key: KeyObject;

  private constructor(key: KeyObject) {
    this.key = key;
  }

  /** Build from the raw 32-byte seed (the `<agent_id>.priv` file contents). */
  static fromSeed(seed: Uint8Array): AgentSigningKey {
    if (seed.length !== KEY_LEN) {
      throw new RangeError(`Ed25519 seed must be 32 bytes, got ${seed.length}`);
    }
    const der = Buffer.concat([PKCS8_PREFIX, Buffer.from(seed)]);
    return new AgentSigningKey(
      createPrivateKey({ key: der, format: "der", type: "pkcs8" }),
    );
  }

  /** Build from a base64 seed (standard or URL-safe, padded or not). */
  static fromBase64(seedB64: string): AgentSigningKey {
    return AgentSigningKey.fromSeed(decodeBase64Any(seedB64));
  }

  /** Generate a fresh key from the OS CSPRNG. */
  static generate(): AgentSigningKey {
    return AgentSigningKey.fromSeed(randomBytes(KEY_LEN));
  }

  /** The raw 32-byte Ed25519 public key. */
  publicKeyBytes(): Buffer {
    const spki = createPublicKey(this.key).export({
      format: "der",
      type: "spki",
    });
    return Buffer.from(spki.subarray(SPKI_PREFIX.length));
  }

  /**
   * URL-safe, unpadded base64 public key.
   *
   * The form `PUT /api/v1/agents/{id}/pubkey` expects in its `pubkey_b64`
   * body field, matching `AgentKeypair::public_base64`.
   */
  publicKeyBase64(): string {
    return this.publicKeyBytes().toString("base64url");
  }

  /** Raw 64-byte Ed25519 signature over `payload`. */
  sign(payload: Buffer): Buffer {
    return cryptoSign(null, payload, this.key);
  }
}

// ---------------------------------------------------------------------------
// #3464 — proof of possession for agent public-key binding
// ---------------------------------------------------------------------------

/**
 * Domain tag for the bind-challenge transcript. Distinct from every other
 * signing domain in the substrate, so a signature minted for a write, a link,
 * a lineage record or a sub-key cert can never be replayed as a possession
 * proof.
 */
export const BIND_CHALLENGE_V1_DOMAIN = "ai-memory/bind-pubkey/v1";

/**
 * The exact bytes a candidate key must sign to prove possession.
 *
 * Domain-separated and LENGTH-PREFIXED, so no field boundary can be shifted to
 * make one transcript read as another (`agentId="a", pubkey="bc"` and
 * `agentId="ab", pubkey="c"` produce different bytes). Mirrors
 * `crate::identity::pubkey_bind::bind_challenge_transcript` byte for byte.
 */
export function bindChallengeTranscript(input: {
  agentId: string;
  pubkeyB64: string;
  nonce: string;
  expiresAt: string;
}): Buffer {
  const parts: Buffer[] = [Buffer.from(BIND_CHALLENGE_V1_DOMAIN, "utf8")];
  for (const field of [input.agentId, input.pubkeyB64, input.nonce, input.expiresAt]) {
    const raw = Buffer.from(field, "utf8");
    const len = Buffer.alloc(4);
    len.writeUInt32BE(raw.length, 0);
    parts.push(len, raw);
  }
  return Buffer.concat(parts);
}

/**
 * Answer a bind challenge; returns the URL-safe-unpadded base64 proof.
 *
 * The public key is derived from `key`, so this can only ever produce a proof
 * for a key the caller actually holds.
 */
export function signBindChallenge(
  key: AgentSigningKey,
  input: { agentId: string; nonce: string; expiresAt: string },
): string {
  const transcript = bindChallengeTranscript({
    agentId: input.agentId,
    pubkeyB64: key.publicKeyBase64(),
    nonce: input.nonce,
    expiresAt: input.expiresAt,
  });
  return key.sign(transcript).toString("base64url");
}

// ---------------------------------------------------------------------------
// Signing
// ---------------------------------------------------------------------------

/** Inputs to {@link signWrite} / {@link attestationFields}. */
export interface AttestationInput {
  agentId: string;
  namespace: string;
  title: string;
  content: string;
  /** Defaults to `"observation"` — the daemon's default when absent. */
  kind?: string;
  /** Defaults to now. Must be the SAME string sent on the wire. */
  createdAt?: string;
}

/**
 * Sign a write envelope and return the STANDARD-base64 signature.
 *
 * Standard (not URL-safe) base64 is what `prepare_signed_store` decodes from
 * the `signature` wire field (`src/identity/attest.rs:194`).
 */
export function signWrite(
  key: AgentSigningKey,
  input: Required<Pick<AttestationInput, "agentId" | "namespace" | "title" | "content" | "kind" | "createdAt">>,
): string {
  const envelope = canonicalCborWrite({
    agentId: input.agentId,
    namespace: input.namespace,
    title: input.title,
    kind: input.kind,
    createdAt: input.createdAt,
    contentSha256: contentSha256(input.content),
  });
  return key.sign(envelope).toString("base64");
}

/** The three wire fields a signed store must carry. */
export interface AttestationFields {
  signature: string;
  created_at: string;
  kind: string;
}

/**
 * Build the `{signature, created_at, kind}` triple for a store request.
 *
 * The three keys drop straight into a `CreateMemoryRequest`. `created_at`
 * defaults to now — the daemon adopts the value verbatim after checking it
 * against the +/-300s freshness window, so it must be the SAME string that was
 * signed, which is why it is returned rather than left to the caller.
 *
 * #3422 — a caller-supplied `createdAt` is folded through
 * {@link canonicalizeCreatedAt} BEFORE signing, so the SDK always signs (and
 * sends) the exact string the daemon will store and later re-derive the
 * signature from on either backend.
 */
export function attestationFields(
  key: AgentSigningKey,
  input: AttestationInput,
): AttestationFields {
  const kind = input.kind ?? "observation";
  const createdAt =
    input.createdAt === undefined ? rfc3339Now() : canonicalizeCreatedAt(input.createdAt);
  return {
    signature: signWrite(key, {
      agentId: input.agentId,
      namespace: input.namespace,
      title: input.title,
      content: input.content,
      kind,
      createdAt,
    }),
    created_at: createdAt,
    kind,
  };
}

File: sdk/python/ai_memory/attestation.py
# Copyright 2026 AlphaOne LLC
# SPDX-License-Identifier: Apache-2.0
"""Ed25519 agent-attestation signing for the HTTP store path (#626 / #2455).

Why this module exists
----------------------

``POST /api/v1/memories`` is classified ``WriteSurface::HttpDirect``, which
**fails closed by default** (``src/identity/attest.rs:130-136`` — the ``_``
arm of ``resolve_require_agent_attestation`` returns ``true`` for
``HttpDirect`` when the env var is unset). An unsigned HTTP store is answered
with ``403 ATTESTATION_FAILED``.

Before #2455 this SDK shipped no signing helper and ``CreateMemory`` carried
no ``signature`` / ``created_at`` fields, so :meth:`AiMemoryClient.store`
could not succeed against a stock daemon at all. This module closes that: it
reproduces the daemon's ``SignableWrite`` envelope byte-for-byte and signs it,
so the SDK works against the **shipped fail-closed default** rather than
requiring the operator to weaken the daemon with
``AI_MEMORY_REQUIRE_AGENT_ATTESTATION=0``.

The envelope
------------

The signature is Ed25519 over the RFC 8949 §4.2.1 deterministic-CBOR encoding
of exactly seven entries (``src/identity/sign.rs::canonical_cbor_write``):

===================  =========================================================
``_dst``             domain-separation tag, always ``"ai-memory/write/v1"``
``kind``             memory kind (e.g. ``"observation"``, ``"decision"``)
``title``            memory title
``agent_id``         the claiming agent id
``namespace``        target namespace
``created_at``       RFC3339 timestamp the caller signed
``content_sha256``   SHA-256 of the content's UTF-8 bytes (CBOR byte string)
===================  =========================================================

Map keys are emitted in *encoded-key* sort order (length first, then
lexicographic), which for this fixed key set is the order shown above. The
byte-level agreement with the Rust encoder is pinned by a fixture printed by
the Rust encoder itself — see ``sdk/fixtures/write_attestation_vector.json``
and ``tests/test_attestation.py``.

Enrolling the key
-----------------

A signature only attests if the daemon has the matching public key bound to
the agent. Bind it once (admin-gated) via
``PUT /api/v1/agents/{id}/pubkey`` with body ``{"pubkey_b64": "..."}`` —
:meth:`AgentSigningKey.public_key_b64` emits the accepted form, and
:meth:`AiMemoryClient.bind_agent_pubkey` calls the route. Without a bound key
the daemon cannot verify and still returns 403 (fail-closed, by design).

Known limits (stated rather than hidden)
----------------------------------------

* The daemon redacts secret-screened content *before* verifying
  (``identity::attest::redact_before_sign``). If your content trips the
  credential screen under ``AI_MEMORY_SECRET_SCREEN_MODE=redact``, the bytes
  the daemon verifies differ from the bytes you signed and the write is
  refused. Under the default ``refuse`` mode such content is rejected earlier
  anyway. Do not put credentials in memories.
* ``created_at`` must be within the daemon's ±300s attestation freshness
  window (``ATTEST_CREATED_AT_SKEW_SECS``); sign shortly before sending.
* ``created_at`` must be the canonical storage-stable rendering (#3422) —
  UTC, ``+00:00``, microsecond-truncated. :func:`canonicalize_created_at`
  produces it and :func:`attestation_fields` applies it for you; the daemon
  refuses anything else, because no other rendering survives a PostgreSQL
  ``TIMESTAMPTZ`` round-trip byte-for-byte.
* The signature commits to those six fields ONLY — not tags, priority, or
  metadata.

Optional dependency
-------------------

Ed25519 needs ``cryptography``; install with ``pip install
"ai-memory-mcp[attestation]"``. Importing this module without it raises a
clear :class:`ImportError` at call time, not at import time, so the rest of
the SDK keeps working.
"""

from __future__ import annotations

import base64
import hashlib
import os
import re
import secrets
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any

__all__ = [
    "DOMAIN_SEPARATION_TAG",
    "AgentSigningKey",
    "attestation_fields",
    "canonical_cbor_write",
    "canonicalize_created_at",
    "content_sha256",
    "rfc3339_now",
    "sign_write",
]

#: The ``_dst`` domain-separation value for the write envelope (#1931).
DOMAIN_SEPARATION_TAG = "ai-memory/write/v1"

#: Length of an Ed25519 seed / public key in bytes.
_KEY_LEN = 32


def _require_cryptography() -> Any:
    try:
        from cryptography.hazmat.primitives.asymmetric import ed25519
    except ImportError as exc:  # pragma: no cover - exercised by env, not CI
        raise ImportError(
            "Ed25519 agent attestation requires the 'cryptography' package. "
            'Install it with: pip install "ai-memory-mcp[attestation]"'
        ) from exc
    return ed25519


# ---------------------------------------------------------------------------
# Deterministic CBOR (RFC 8949 §4.2.1) — only the shapes this envelope needs.
# ---------------------------------------------------------------------------


def _cbor_head(major: int, length: int) -> bytes:
    """Encode a CBOR head for ``major`` with the minimal-width length."""
    prefix = major << 5
    if length < 24:
        return bytes([prefix | length])
    if length < 0x100:
        return bytes([prefix | 24, length])
    if length < 0x10000:
        return bytes([prefix | 25]) + length.to_bytes(2, "big")
    if length < 0x1_0000_0000:
        return bytes([prefix | 26]) + length.to_bytes(4, "big")
    return bytes([prefix | 27]) + length.to_bytes(8, "big")


def _cbor_text(value: str) -> bytes:
    raw = value.encode("utf-8")
    return _cbor_head(3, len(raw)) + raw


def _cbor_bytes(value: bytes) -> bytes:
    return _cbor_head(2, len(value)) + value


def content_sha256(content: str) -> bytes:
    """SHA-256 over the UTF-8 bytes of ``content`` (the bounded commitment)."""
    return hashlib.sha256(content.encode("utf-8")).digest()


def canonical_cbor_write(
    *,
    agent_id: str,
    namespace: str,
    title: str,
    kind: str,
    created_at: str,
    content_hash: bytes,
) -> bytes:
    """Deterministic-CBOR encoding of the ``SignableWrite`` envelope.

    Byte-identical to ``crate::identity::sign::canonical_cbor_write``.

    Args:
        content_hash: 32-byte SHA-256 digest from :func:`content_sha256`.

    Raises:
        ValueError: if ``content_hash`` is not exactly 32 bytes.
    """
    if len(content_hash) != _KEY_LEN:
        raise ValueError(
            f"content_hash must be a 32-byte SHA-256 digest, got {len(content_hash)}"
        )
    # RFC 8949 §4.2.1 orders map keys by their ENCODED bytes: shorter keys
    # first, then bytewise. For this fixed key set that is the literal order
    # below ('_' 0x5F sorts before 'k' 0x6B at length 4). Hard-coding the
    # resolved order keeps the encoder allocation-free and, more importantly,
    # makes any future key addition an explicit, reviewable re-ordering rather
    # than a silent one. The order is pinned against the Rust-emitted fixture.
    entries = (
        ("_dst", _cbor_text(DOMAIN_SEPARATION_TAG)),
        ("kind", _cbor_text(kind)),
        ("title", _cbor_text(title)),
        ("agent_id", _cbor_text(agent_id)),
        ("namespace", _cbor_text(namespace)),
        ("created_at", _cbor_text(created_at)),
        ("content_sha256", _cbor_bytes(content_hash)),
    )
    out = bytearray(_cbor_head(5, len(entries)))
    for key, encoded_value in entries:
        out += _cbor_text(key)
        out += encoded_value
    return bytes(out)


#: RFC3339 date-time, decomposed so sub-second digits are handled explicitly.
_RFC3339_PATTERN = re.compile(
    r"^(\d{4})-(\d{2})-(\d{2})[Tt ](\d{2}):(\d{2}):(\d{2})"
    r"(?:\.(\d+))?([Zz]|[+-]\d{2}:\d{2})$"
)


def canonicalize_created_at(value: str) -> str:
    """Canonicalize an RFC3339 timestamp into the ONE ``created_at`` rendering
    both daemon storage backends return byte-for-byte (#3422).

    The daemon REFUSES to attest any other rendering. ``created_at`` sits inside
    the signed ``SignableWrite`` envelope as TEXT, and every re-verification
    (the store gate, and the federation receive path on a relayed row) rebuilds
    the envelope from the *persisted row*: SQLite returns the stored TEXT
    verbatim, while PostgreSQL stores ``TIMESTAMPTZ`` (microseconds) and
    re-renders the readback as ``+00:00``. A signature minted over ``"...Z"``,
    over a non-UTC offset, or over nanosecond precision could therefore never
    be re-derived from a PostgreSQL row.

    Canonical form: UTC, ``+00:00`` offset, microseconds TRUNCATED (never
    rounded — that is what ``TIMESTAMPTZ`` does), and 0, 3 or 6 fractional
    digits, the shortest width that represents the value exactly (chrono's
    ``SecondsFormat::AutoSi``).

    Raises:
        ValueError: if ``value`` is not an RFC3339 date-time.
    """
    match = _RFC3339_PATTERN.match(value.strip())
    if match is None:
        raise ValueError(f"created_at must be an RFC3339 date-time, got: {value!r}")
    year, month, day, hour, minute, second, fraction, offset = match.groups()
    # TIMESTAMPTZ keeps microseconds and drops the rest — truncate, never round.
    micros = int(f"{fraction or ''}000000"[:6])
    tz = timezone.utc if offset in ("Z", "z") else timezone(_offset_delta(offset))
    moment = datetime(
        int(year),
        int(month),
        int(day),
        int(hour),
        int(minute),
        int(second),
        micros,
        tzinfo=tz,
    ).astimezone(timezone.utc)
    # Explicit field formatting rather than strftime: %Y is not guaranteed to
    # zero-pad (or even accept) years below 1000 across platforms.
    whole = (
        f"{moment.year:04d}-{moment.month:02d}-{moment.day:02d}"
        f"T{moment.hour:02d}:{moment.minute:02d}:{moment.second:02d}"
    )
    if moment.microsecond == 0:
        frac = ""
    elif moment.microsecond % 1000 == 0:
        frac = f".{moment.microsecond // 1000:03d}"
    else:
        frac = f".{moment.microsecond:06d}"
    return f"{whole}{frac}+00:00"


def _offset_delta(offset: str) -> timedelta:
    """``+HH:MM`` / ``-HH:MM`` as a :class:`datetime.timedelta`."""
    sign = -1 if offset.startswith("-") else 1
    return sign * timedelta(hours=int(offset[1:3]), minutes=int(offset[4:6]))


def rfc3339_now() -> str:
    """Current UTC time in the canonical ``created_at`` form the daemon
    persists and re-derives signatures from (#3422) — UTC, ``+00:00``,
    microsecond-truncated.
    """
    return canonicalize_created_at(datetime.now(timezone.utc).isoformat())


# ---------------------------------------------------------------------------
# Key handling
# ---------------------------------------------------------------------------


class AgentSigningKey:
    """An Ed25519 signing key for agent write attestation.

    Wraps a 32-byte seed — the exact on-disk format the daemon's own key
    store uses (``<key_dir>/<agent_id>.priv``, 32 raw bytes, mode 0600), so a
    key generated by ``ai-memory identity generate`` can be loaded directly
    with :meth:`from_file`.
    """

    __slots__ = ("_private",)

    def __init__(self, seed: bytes) -> None:
        if len(seed) != _KEY_LEN:
            raise ValueError(f"Ed25519 seed must be 32 bytes, got {len(seed)}")
        ed25519 = _require_cryptography()
        self._private = ed25519.Ed25519PrivateKey.from_private_bytes(seed)

    @classmethod
    def generate(cls) -> AgentSigningKey:
        """Generate a fresh key from the OS CSPRNG."""
        return cls(secrets.token_bytes(_KEY_LEN))

    @classmethod
    def from_file(cls, path: str | os.PathLike[str]) -> AgentSigningKey:
        """Load a 32-raw-byte ``<agent_id>.priv`` file."""
        return cls(Path(path).read_bytes())

    @classmethod
    def from_base64(cls, seed_b64: str) -> AgentSigningKey:
        """Load a seed from base64 (standard or URL-safe, padded or not)."""
        return cls(_b64_decode_any(seed_b64))

    def seed_bytes(self) -> bytes:
        """The raw 32-byte seed — write it mode 0600 if you persist it."""
        from cryptography.hazmat.primitives import serialization

        return self._private.private_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PrivateFormat.Raw,
            encryption_algorithm=serialization.NoEncryption(),
        )

    def public_key_bytes(self) -> bytes:
        """The raw 32-byte Ed25519 public key."""
        from cryptography.hazmat.primitives import serialization

        return self._private.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )

    def public_key_b64(self) -> str:
        """URL-safe, unpadded base64 public key.

        This is the form ``PUT /api/v1/agents/{id}/pubkey`` expects in its
        ``pubkey_b64`` body field, matching ``AgentKeypair::public_base64``.
        """
        return base64.urlsafe_b64encode(self.public_key_bytes()).decode("ascii").rstrip("=")

    def sign(self, payload: bytes) -> bytes:
        """Raw 64-byte Ed25519 signature over ``payload``."""
        return self._private.sign(payload)


def _b64_decode_any(value: str) -> bytes:
    """Decode standard or URL-safe base64, with or without padding."""
    raw = value.strip()
    padded = raw + "=" * (-len(raw) % 4)
    try:
        return base64.urlsafe_b64decode(padded)
    except (ValueError, TypeError):
        return base64.b64decode(padded)


# ---------------------------------------------------------------------------
# Signing
# ---------------------------------------------------------------------------


def sign_write(
    key: AgentSigningKey,
    *,
    agent_id: str,
    namespace: str,
    title: str,
    kind: str,
    created_at: str,
    content: str,
) -> str:
    """Sign a write envelope and return the STANDARD-base64 signature.

    Standard (not URL-safe) base64 is what ``prepare_signed_store`` decodes
    from the ``signature`` wire field (``src/identity/attest.rs:194``).
    """
    envelope = canonical_cbor_write(
        agent_id=agent_id,
        namespace=namespace,
        title=title,
        kind=kind,
        created_at=created_at,
        content_hash=content_sha256(content),
    )
    return base64.b64encode(key.sign(envelope)).decode("ascii")


def attestation_fields(
    key: AgentSigningKey,
    *,
    agent_id: str,
    namespace: str,
    title: str,
    content: str,
    kind: str = "observation",
    created_at: str | None = None,
) -> dict[str, str]:
    """Build the ``{"signature", "created_at", "kind"}`` triple for a store.

    The three keys drop straight into a ``CreateMemory`` body. ``created_at``
    defaults to now — the daemon adopts the value verbatim after checking it
    against the ±300s freshness window, so it must be the SAME string that
    was signed, which is why it is returned rather than left to the caller.

    #3422 — a caller-supplied ``created_at`` is folded through
    :func:`canonicalize_created_at` BEFORE signing, so the SDK always signs
    (and sends) the exact string the daemon will store and later re-derive the
    signature from on either backend.

    ``kind`` defaults to ``"observation"``, the daemon's own default when the
    field is absent — so an unspecified kind signs what the server will store.
    """
    ts = rfc3339_now() if created_at is None else canonicalize_created_at(created_at)
    return {
        "signature": sign_write(
            key,
            agent_id=agent_id,
            namespace=namespace,
            title=title,
            kind=kind,
            created_at=ts,
            content=content,
        ),
        "created_at": ts,
        "kind": kind,
    }
